src/ng/compile.js

```diff
@@ -2807,53 +2803,69 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           }
         }
 
-        // Initialize bindToController bindings
-        for (var name in elementControllers) {
-          var controllerDirective = controllerDirectives[name];
-          var controller = elementControllers[name];
-          var bindings = controllerDirective.$$bindings.bindToController;
+        if (controllerDirectives) {
+          elementControllers = createMap();
+          for (var name in controllerDirectives) {
+            var directive = controllerDirectives[name];
+            var locals = {
+              $scope: directive === newIsolateScopeDirective || directive.$$isolateScope ? isolateScope : scope,
+              $element: $element,
+              $attrs: attrs,
+              $transclude: transcludeFn
+            };
 
-          controller.instance = controller();
-          $element.data('$' + controllerDirective.name + 'Controller', controller.instance);
-          controller.bindingInfo =
-            initializeDirectiveBindings(controllerScope, attrs, controller.instance, bindings, controllerDirective);
-          }
+            var controllerConstructor = directive.controller;
+            if (controllerConstructor === '@') {
+              controllerConstructor = attrs[name];
+            }
 
-        // Bind the required controllers to the controller, if `require` is an object and `bindToController` is truthy
-        forEach(controllerDirectives, function(controllerDirective, name) {
-          var require = controllerDirective.require;
-          if (controllerDirective.bindToController && !isArray(require) && isObject(require)) {
-            extend(elementControllers[name].instance, getControllers(name, require, $element, elementControllers));
+            var instance = $controller(controllerConstructor, locals, false, directive.controllerAs);
```

Comment:
One difference with the previous implementation is that all controller instances where available on the element before instantiating a controller. (Now it depends on the relative order of controllers.)
But I think this is fine since it is not recommended to access controllers directly off of `$element`, but require them and access them in `ngOnInit()`.

Reply:
'Instances available before instantiation` is exactly what we're getting rid of.

Reply:
Not really :grin:

First of all, this is a refactoring commit, so we shouldn't be getting rid of anything 😛
The associated, feature-removing PR got rid of bindings (and instances?) being assigned to the controller instance before instantiation. But if someone tried to grab the instances with `$element.controller('xyz')`, they would be there (possibly uninstantiated, but that doesn't matter).

So, one could argue that this commit introduces a breaking change. But like I said, _"I think this is fine since it is not recommended to access controllers directly off of `$element`, but one should require them and access them in `ngOnInit()`"_. :smiley:

Reply:
> they would be there (possibly uninstantiated, but that doesn't matter)

And only if the class isn't an ES6 class.